### PR TITLE
AirfRANS: vol-weight warm-up schedule (1x→10x linear ramp)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Current AF champion uses a fixed `--vol-loss-weight 10.0` from epoch 1. But early in training when predictions are poor, the volume loss gradient is extremely noisy — upweighting it 10x from the start may corrupt early surface learning and slow convergence to the surface basin that EMA then stabilizes.

**Proposed fix:** Implement a linear warm-up of vol-loss-weight from 1.0 to 10.0 over the first 200 epochs, then hold at 10.0 for the remainder. This lets the model first learn reasonable surface representations (where the signal-to-noise ratio is higher), then gradually shift focus to volume predictions once the model has a stable foundation.

**Why this is different from chihiro #3226 (20x→10x decay):** Chihiro tests *decreasing* vol-weight (front-loading volume attention then relaxing). This experiment tests the *opposite direction* — starting gentle and ramping up. These are complementary hypotheses about whether early or late volume emphasis is more effective.

**Physics motivation:** The surface boundary layer must be resolved first before volume predictions can be accurate (volume fields are physically determined by surface conditions via the Navier-Stokes equations). Training the surface first, then shifting to volume, follows the physical causality.

## Instructions

Modify `target/icml2026/train.py` to implement dynamic vol-weight scheduling:

1. Add a `--vol-weight-warmup-epochs` argument (default: 0 = disabled, i.e., fixed vol-weight as before).

2. When `--vol-weight-warmup-epochs N` is set, linearly interpolate `vol_loss_weight` from 1.0 to the target `--vol-loss-weight` value over the first N epochs:
   ```python
   if epoch < vol_weight_warmup_epochs:
       current_vol_weight = 1.0 + (vol_loss_weight - 1.0) * (epoch / vol_weight_warmup_epochs)
   else:
       current_vol_weight = vol_loss_weight
   ```

3. Log the current `vol_loss_weight` to W&B each epoch so we can verify the schedule.

4. Run 3 trials:

**Trial 1 — 200-epoch warm-up (recommended):**
```
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0 \
  --vol-weight-warmup-epochs 200 --wandb_group af-vol-warmup
```

**Trial 2 — 100-epoch warm-up (faster ramp):**
Same as Trial 1 but `--vol-weight-warmup-epochs 100`

**Trial 3 — 50-epoch warm-up (aggressive ramp):**
Same as Trial 1 but `--vol-weight-warmup-epochs 50`

All other hyperparameters must match the champion config exactly.

## Baseline

Current AF best (PR #3135):
- **val surface_mse:** 0.000296 at epoch 704
- **val vol_mse:** 0.002039 at epoch 743
- **Goal:** vol_mse < 0.0017 without regressing surface_mse above 0.000296
- **W&B run:** sh2zyfwr
- **Reproduce:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999 --vol-loss-weight 10.0`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT nezuko:

## Results

### Summary

**Hypothesis not supported.** None of the warmup schedules beat the baseline on either surface_mse or volume_mse. Two out of three trials experienced catastrophic collapse during training. The fixed 10x vol-weight from epoch 1 (baseline) remains superior.

### Metrics

| Trial | Warmup | Best Val Surface MSE | Best Val Vol MSE | Test Surface MSE | Test Vol MSE | Best Epoch | Total Epochs | Status |
|-------|--------|---------------------|------------------|------------------|--------------|------------|--------------|--------|
| **Baseline** | none (fixed 10x) | **0.000296** (ep704) | **0.002039** (ep743) | — | — | — | ~770 | stable |
| T1 (warmup 200) | 200 ep | 0.000421 (ep566) | 0.002719 (ep546) | 0.000774 | 0.004236 | 566 | 766 | **collapsed ~ep600** |
| T2 (warmup 100) | 100 ep | 0.000379 (ep571) | 0.002458 (ep684) | 0.001142 | 0.004047 | 571 | 755 | stable |
| T3 (warmup 50) | 50 ep | 0.001098 (ep166) | 0.008497 (ep228) | 0.001505 | 0.012571 | 166 | 759 | **collapsed ~ep200** |

### vs Baseline

- **T2 (warmup 100):** Best performer. Val surface_mse 28% above baseline (0.000379 vs 0.000296). Val vol_mse 21% above baseline (0.002458 vs 0.002039). Only stable trial.
- **T1 (warmup 200):** Val surface_mse 42% above baseline. Collapsed into degenerate predictions (constant 0.621 surface_mse) around epoch 600. Best checkpoint preserved at ep566.
- **T3 (warmup 50):** Val surface_mse 271% above baseline. Collapsed earliest around epoch 200. Best checkpoint at ep166 (only 3.3x the ramp length before collapse).

### Stability Analysis

The most striking finding is the catastrophic instability. Two of three warmup trials collapsed into degenerate states where the model produced constant, near-zero predictions for Ux/Uy/nut components and a fixed ~2.48 for pressure:
- **T1 (warmup 200):** Collapsed ~ep600 with identical degenerate outputs every epoch (surface_mse_Ux=0.0, Uy=0.0, nut=0.0, p=2.483)
- **T3 (warmup 50):** Collapsed ~ep200 with similar degenerate pattern (surface_mse ~0.57)
- **T2 (warmup 100):** Survived but showed oscillation around 0.0005–0.0008, never improving past ep571

The collapse pattern correlates with cosine LR restarts (T_max=50). After reaching full 10x weight, the cosine restart resets LR to 6e-4 against the now-heavy volume loss term, creating destructive gradient spikes. The baseline (which starts at 10x from epoch 1) presumably adapts early optimizer state to handle these restarts, while the warmup trials reach full weight with optimizer state tuned to lighter weights.

### Commands Used

```bash
# Trial 1 (warmup 200)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --ema-decay 0.999 --volume-loss-weight 10.0 \
  --vol-weight-warmup-epochs 200 --agent nezuko --wandb-name "nezuko/af-warmup200" --wandb-group af-vol-warmup

# Trial 2 (warmup 100)
# Same with --vol-weight-warmup-epochs 100 --wandb-name "nezuko/af-warmup100" on GPU 1

# Trial 3 (warmup 50)
# Same with --vol-weight-warmup-epochs 50 --wandb-name "nezuko/af-warmup50" on GPU 2
```

### W&B Runs
- T1 (warmup 200): `9tc2udwf` — nezuko/af-warmup200
- T2 (warmup 100): `gp149j2g` — nezuko/af-warmup100
- T3 (warmup 50): `o8ifqyac` — nezuko/af-warmup50
- Group: `af-vol-warmup`
- Entity/project: `wandb-applied-ai-team/senpai-v1`

### Peak Memory
~42 GB VRAM per trial (same as baseline — warmup does not change memory footprint)

### What Happened

The hypothesis that "surface first, then volume" training would improve convergence was not supported. The warmup scheduling:

1. **Did not improve surface learning** — the delay in volume weight didn't produce meaningfully better surface representations during the low-weight phase
2. **Introduced catastrophic instability** — the transition to full weight created a regime change that the optimizer (and cosine LR restarts) could not handle, causing 2/3 trials to collapse
3. **The "Goldilocks" warmup (100 ep) survived** but still underperformed the baseline by 28% on surface and 21% on volume

The baseline's approach of starting at full 10x weight from epoch 1 appears to be the correct choice — it allows the optimizer state to adapt from the beginning to the combined loss landscape, avoiding the dangerous weight transition.

### Suggested Follow-ups

1. **Warmup + LR warmup coupling:** The core issue may be that cosine LR restarts interact badly with a changing loss weight. A warmup that also adjusts the LR restart schedule (e.g., longer T_max during warmup, or no restarts during warmup) might avoid the instability.
2. **Gentler ramp target:** Instead of 1x→10x, try 5x→10x (halving the gradient ratio change). This smaller shift might avoid triggering collapse while still gaining any benefit from gentler early training.
3. **Exponential warmup instead of linear:** An exponential ramp spends more time near the target value, reducing the rate of change when the weight is highest and most impactful.
4. **Volume-weight decay (complementary to Chihiro #3226):** The opposite direction — starting heavy and decaying — may actually be more compatible with cosine restarts since the optimizer adapts early to the larger weight.

**Comment by tcapelle** [comment]
Closing — no trial beat baseline on either surface_mse or vol_mse, and 2/3 trials collapsed catastrophically.

**W&B-verified metrics (best checkpoint):**
| Trial | Warm-up | surface_mse | vs baseline | vol_mse | vs baseline | Status |
|-------|---------|-------------|-------------|---------|-------------|--------|
| T1 | 200 ep | 0.000421 | +42.3% | 0.003456 | +69.5% | COLLAPSED (ratio 1473x) |
| T2 | 100 ep | 0.000379 | +28.1% | 0.003443 | +68.8% | stable |
| T3 | 50 ep | 0.001098 | +270.9% | 0.010298 | +405.1% | COLLAPSED (ratio 536x) |

Baseline: surface_mse=0.000296, vol_mse=0.002039

**Why closing vs send-back:**
The collapse signature in T1/T3 (constant predictions: Ux=Uy=nut≈0, p≈2.48) is caused by cosine LR restarts (T_max=50) resetting LR to 6e-4 at the moment vol_weight reaches 10x, creating destructive gradient spikes. T2 is stable but +28% surface / +69% volume — too far from baseline to justify iteration.

The core hypothesis (surface-first warmup improves final convergence) is not supported. The champion config's constant vol_weight=10x from epoch 0 remains superior. The vol_mse target of <0.0017 was not approached (best: 0.003443, 2x the target).

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3232 because ADVISOR: Closing — vol-weight warm-up schedule is dead for AF. 2 of 3 trials collapsed (T1 200ep ramp: 1473x grad explosion, T3 50ep: 536x explosion). Best stable trial (T2 100ep): surface +28%, vol +69% — not close. Cosine LR restarts interact destructively with the weight ramp — the LR peak injects instability before vol-weight has reached its stabilizing 10x level. Static 10x from epoch 0 is now triple-confirmed as optimal (this + curriculum #3226 + vol>10x #3204)..
